lib/amazonka/src/Amazonka/Env/Hooks.hs

```diff
@@ -0,0 +1,420 @@
+  Hooks ->
+  Hooks
+waitHook f hooks@Hooks {wait} =
+  hooks {wait = f wait}
+
+{-# INLINE configuredRequestHook #-}
+configuredRequestHook ::
+  ( forall a.
+    (AWSRequest a, Typeable a) =>
+    Hook (Request a) ->
+    Hook (Request a)
+  ) ->
+  Hooks ->
+  Hooks
+configuredRequestHook f hooks@Hooks {configuredRequest} =
+  hooks {configuredRequest = f configuredRequest}
+
+{-# INLINE signedRequestHook #-}
+signedRequestHook ::
+  ( forall a.
+    (AWSRequest a, Typeable a) =>
+    Hook_ (Signed a) ->
+    Hook_ (Signed a)
+  ) ->
+  Hooks ->
+  Hooks
+signedRequestHook f hooks@Hooks {signedRequest} =
+  hooks {signedRequest = f signedRequest}
+
+{-# INLINE clientRequestHook #-}
+clientRequestHook ::
+  (Hook ClientRequest -> Hook ClientRequest) ->
+  Hooks ->
+  Hooks
+clientRequestHook f hooks@Hooks {clientRequest} =
+  hooks {clientRequest = f clientRequest}
+
+{-# INLINE clientResponseHook #-}
+clientResponseHook ::
+  ( forall a.
+    (AWSRequest a, Typeable a) =>
+    Hook_ (Request a, ClientResponse ()) ->
+    Hook_ (Request a, ClientResponse ())
+  ) ->
+  Hooks ->
+  Hooks
+clientResponseHook f hooks@Hooks {clientResponse} =
+  hooks {clientResponse = f clientResponse}
+
+{-# INLINE requestRetryHook #-}
+requestRetryHook ::
+  ( forall a.
+    (AWSRequest a, Typeable a) =>
+    Hook_ (Request a, Text, Retry.RetryStatus) ->
+    Hook_ (Request a, Text, Retry.RetryStatus)
+  ) ->
+  Hooks ->
+  Hooks
+requestRetryHook f hooks@Hooks {requestRetry} =
+  hooks {requestRetry = f requestRetry}
+
+{-# INLINE awaitRetryHook #-}
+awaitRetryHook ::
+  ( forall a.
+    (AWSRequest a, Typeable a) =>
+    Hook_ (Request a, Wait a, Accept, Retry.RetryStatus) ->
+    Hook_ (Request a, Wait a, Accept, Retry.RetryStatus)
+  ) ->
+  Hooks ->
+  Hooks
+awaitRetryHook f hooks@Hooks {awaitRetry} =
+  hooks {awaitRetry = f awaitRetry}
+
+{-# INLINE responseHook #-}
+responseHook ::
+  ( forall a.
+    (AWSRequest a, Typeable a) =>
+    Hook_ (Request a, ClientResponse (AWSResponse a)) ->
+    Hook_ (Request a, ClientResponse (AWSResponse a))
+  ) ->
+  Hooks ->
+  Hooks
+responseHook f hooks@Hooks {response} =
+  hooks {response = f response}
+
+{-# INLINE errorHook #-}
+errorHook ::
+  ( forall a.
+    (AWSRequest a, Typeable a) =>
+    Hook_ (Request a, Error) ->
+    Hook_ (Request a, Error)
+  ) ->
+  Hooks ->
+  Hooks
+errorHook f hooks@Hooks {error} =
+  hooks {error = f error}
+
+-- | Turn a @'Hook' a@ into another @'Hook' a@ that does nothing.
+--
+-- -- Example: remove all request hooks:
+-- @
+-- requestHook noHook :: Hooks -> Hooks
+-- @
+noHook :: Hook a -> Hook a
+noHook _ _ = pure
+
+-- | Turn a @'Hook_' a@ into another @'Hook_' a@ that does nothing.
+--
+-- -- Example: Remove all response hooks:
+-- @
+-- responseHook noHook_ :: Hooks -> Hooks
+-- @
+noHook_ :: Hook_ a -> Hook_ a
+noHook_ _ _ _ = pure ()
+
+-- | When @a@ and @b@ are the same type, run the given 'Hook a' after
+-- all others, otherwise do nothing.
```

Comment:
i had a bit of trouble understanding what "do nothing" meant (since it's just after "run the given Hook a", at first i understood the "do nothing" to be when the hooks are run, not when this function is called).

Maybe "Append the provided 'Hook a' to the existing hooks when 'a' and 'b' match. Otherwise, leave existing hooks unaffected"?

Reply:
Your suggestion sounds to me like it decides whether to append a hook before the time hooks are run. I'll try a different wording, let me know if you like it.

Reply:
Updated and pushed. How do you like the docs now?

Reply:
neat, thanks!